Add New VRAM Example

bloodeagle40234 opened the pull request:
This commit provides a Python example for VRAM (GPU memory) transfer. The block-based transfer model is inspired by vLLM and demonstrates transfer bandwidth under specific configurations.

## What?

This commit adds a new Python example for VRAM (GPU memory) transfer. The example follows vLLM's block-based memory management and shows how to configure NIXL using the `make_prepped_xfer` API.

## Why?

The current NIXL upstream does not include a VRAM-based example, which may make it unclear for users how to configure and debug data transfers between GPUs. To address this, this PR provides a working example. It also demonstrates index-based transfer, as used in vLLM, along with bandwidth performance results. This helps users understand KV cache transfer in disaggregated inference scenarios.

## How?

This PR includes three components and one documentation file. The example is based on a server-client model: the client reads specific tensor blocks from the running server and outputs performance results to the console.

bloodeagle40234 committed:
Add New VRAM Example

This commit provides VRAM transfer example in Python.
The block base transfer model is inspired by vLLM and show
transfer bandwidth with the specific configurations.

Signed-off-by: Kota Tsuyuzaki <bloodeagle40234@gmail.com>

bloodeagle40234 committed:
Fix Copyright Identifier

Because copyright check apparently requires NVIDIA's copyright.

https://github.com/ai-dynamo/nixl/blob/main/.github/workflows/copyright-check.sh#L32-L35

bloodeagle40234 committed:
Move check_remote_metadata call only once per connection

mkhazraee commented:
Hi Kota,

Really sorry for the delayed reply, I was giving a talk at the PyTorch conference, then some work related to GTC, then attending Ray Summit, and I lost track of this PR.

- Regarding using tensors, I see you do: 
tensors = [torch.zeros(tensor_size, dtype=torch.bfloat16, device=device) for x in range(args.layers)] 
First of all, you can already pass a list of tensors directly to NIXL API, and it would get the addr/len/devID parts. No need to extract them yourself. My other point was good chance these sensors end up back to back in memory, so you can make them a single tensor, and then have blocks within it as individual tensor. So let's say you wanted 2D tensors, you can define a single 3D tensor, use the single tensor for registration, and then use slicing during prep. That becomes more similar to vLLM. That being said, that's what they're doing per layer, different layer tensors are not necessarily back to back, so this was a suggestion based on your current code.
   
- Regarding notifications, yes you can put whatever string (actually byte array) you want. That being said, unless you block between transfers (which you're due to while 1), there is no guarantee of ordering between transfers. Let's say you make 10 transfers, 9 of them being "KEEPALIVE" and one being "COMPLETE", and you start them in non-blocking manner, then the transfer with COMPLETE notification might finish before an earlier KEEPALIVE. So I think this might be a tad confusing as an example. Maybe you can put a transfer number (like 1 to 10) as the notification, and wait until you receive all of them. (similar to vLLM).

- Regarding asynchronous aspect, there are two levels of async, one being within each xfer_req, and one across xfer_req. If you want to only benefit from the internal one, and have a blocking mechanism across xfer_req (to make them ordered for example), that's fine to do what you're doing in that aspect, but let's add a comment and explanation in README, this is how to enforce ordering across Xfers as an example, otherwise they could've been started together and based on different notification messages understand when all the transfers are done.

Moein

bloodeagle40234 committed:
Change tensor alignment

Original code creates contiguous tensors for each layers but
it may cause divided alignments among tensors and make the code
complicated on the NIXL memory registration.

This commit simplifies the tensor allocation with just a single
contiguous tensor. Then, use leyer index to access each layer
tensor for block index calculation. This also simplifies the
NIXL memory registration just to pass the tensors for get_reg_descs
method.

bloodeagle40234 committed:
Use sequence number instead of status messages

Older code uses status messages either KEEPALIVE or COMPLETE
to confirm the server/client communication status for the transfer
cycle. However, as Moein said in the review comment, the status
message would be flipped with asynchronous manner. Avoiding the
missing such a status change, newer code introduces to use sequence
number for the asynchronous manner. With this change, new protocol
between the client and the server is like:

- a client sends SYN with total number of sequences
- a server checks the sequence number of each tranfer
- then the server will exit all sequence completely arrived

bonus:
with this manner, we don't have to wait UUID signal base
check_remote_xfer_done blocking call. We just receive the
notification and assert all transfer done, instead.

bloodeagle40234 commented:
Thanks @tstamler for progressing this PR.

Is there anything I can do to help with the CI failure?
It seems to be related to NVIDIA's internal CI process, so I couldn't find any details about the error for now.

mkhazraee commented:
Thanks Kota for the reminder, I lost track of this. 

Nitpicks first: the first commit to use a single tensor is what I meant, only minor thing is when you want to register a tensor, you don't need to pass the memory type. It's a tensor, NIXL python wrapper will figure it internally. In second commit, I see again VRAM is passed to prepXferDlist, but if it's already a dlist, then the type is included, that VRAM will be ignored. I see pycache is included in the files, let's remove it. 

Also let's clarify what makes this example unique now, to name it based on that. I think these 2 examples are the closest ones:
https://github.com/ai-dynamo/nixl/blob/main/examples/python/basic_two_peers.py
https://github.com/ai-dynamo/nixl/blob/main/examples/python/expanded_two_peers.py

Then we can come up with a better name. (I think it might need one more round of clean up, it's a tad hard to follow the full logic of the example, but first I want to know what is the core of challenges you had to figure and clean it up accordingly).

Sorry again for the delays and thanks for your contribution,
Moein

bloodeagle40234 committed:
Remove unnecessary VRAM args and unused commentout
